File: src/gh.ts
import { spawnSync } from "node:child_process";

function getGhBin(): string {
  return process.env.GCL_GH_BIN || "gh";
}

export interface GhResult {
  status: number;
  stdout: string;
  stderr: string;
}

export function runGh(args: string[], options: Parameters<typeof spawnSync>[2] = {}): GhResult {
  const res = spawnSync(getGhBin(), args, {
    encoding: "utf8",
    env: process.env,
    ...options
  });

  const stdout = typeof res.stdout === "string" ? res.stdout : String(res.stdout || "");
  const stderr = typeof res.stderr === "string" ? res.stderr : String(res.stderr || "");

  return {
    status: res.status ?? 1,
    stdout,
    stderr
  };
}

export function ghAuthStatus(): GhResult {
  return runGh(["auth", "status", "-h", "github.com"]);
}

export function ghApi(args: string[]): string {
  const res = runGh(["api", ...args]);
  if (res.status !== 0) {
    const err = new Error(`gh api failed: ${res.stderr || res.stdout}`.trim()) as Error & { code?: number };
    err.code = res.status;
    throw err;
  }
  return res.stdout;
}

File: internal/gh/gh.go
package gh

import (
	"fmt"
	"os"
	"os/exec"
)

func bin() string {
	if v := os.Getenv("GCL_GH_BIN"); v != "" {
		return v
	}
	return "gh"
}

func Run(args ...string) (string, string, int) {
	cmd := exec.Command(bin(), args...)
	cmd.Env = os.Environ()
	stdout, err := cmd.Output()
	if err == nil {
		return string(stdout), "", 0
	}
	if ee, ok := err.(*exec.ExitError); ok {
		return string(stdout), string(ee.Stderr), ee.ExitCode()
	}
	return "", err.Error(), 1
}

func AuthStatus() (bool, string) {
	out, errOut, code := Run("auth", "status", "-h", "github.com")
	if code == 0 {
		return true, out
	}
	if errOut != "" {
		return false, errOut
	}
	return false, out
}

func API(args ...string) (string, error) {
	all := append([]string{"api"}, args...)
	out, errOut, code := Run(all...)
	if code != 0 {
		return "", fmt.Errorf("gh api failed: %s", firstNonEmpty(errOut, out))
	}
	return out, nil
}

func firstNonEmpty(a, b string) string {
	if a != "" {
		return a
	}
	return b
}
